### PR TITLE
[go_router] Fixes an issue where the params are removed after popping

### PR DESCRIPTION
**Pull request by chunhtai**
pop will remove pathparameter from routeMatchList, and the goRouterState also use the same reference to the pathparameter in the routeMatchList, thus parameters got removed as well

## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [ ] I signed the [CLA].
- [ ] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [ ] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

File: packages/go_router/test/go_router_state_test.dart [hunk 1]
```diff
@@ -67,6 +67,40 @@ void main() {
...
+            ]),
+      ];
+      final GoRouter router = await createRouter(routes, tester);
+      await tester.pumpAndSettle();
+      expect(find.text('1 /'), findsOneWidget);
+
+      router.go('/123');
+      await tester.pumpAndSettle();
+      expect(find.text('2 123'), findsOneWidget);
+      // The query parameter is removed, so is the location in first page.
```
Comment (hangyujin) [inline packages/go_router/test/go_router_state_test.dart:98]: What does this comment mean here?  which query parameter and location in first page is removed?

File: packages/go_router/test/go_router_state_test.dart [hunk 2]
```diff
@@ -67,6 +67,40 @@ void main() {
...
+      router.pop();
+      await tester.pump();
+      expect(find.text('2 123'), findsOneWidget);
```
Comment (hangyujin) [inline packages/go_router/test/go_router_state_test.dart:101]: I don't get why it's still at page 2 after poping.

Comment (chunhtai) [reply packages/go_router/test/go_router_state_test.dart:102]: Added a comment to explain.

**Review by hangyujin — APPROVED** [review]
LGTM

**Pull request merged**